feat: Windows support

rllyy97 opened the pull request:
## Main Changes

Added Windows build support!
Builds with the NSIS windows target.

Adjustments for Windows:
- Filepaths
  - Some replace / join fixes to split by '/' and '\', and join using the proper character
- Title bar buttons
  - Had to make some changes with the right aligned windows control buttons
- No transparency (only for Windows)
  - Electron doesn't natively have the blur that Macs have, so it's very hard to see certain content if it's transparent
  - There are packages out there to get the blur but it looks like it has decent performance implications
- Dependency patches
  - The unmaintained `pdf-parse` package that is used in the `llamaindex` dependency has a bug that crashes the app when built
    - https://gitlab.com/autokent/pdf-parse/-/issues/24
  - It does not look like this will be fixed so I've patched the package locally as a workaround

I'm not exactly sure how you go about publishing the packaged files, but just to test I was running `package:win` and then running the resulting setup .exe file and it all seemed to work great

## Screenshots

![image](https://github.com/UdaraJay/Pile/assets/25409734/2467bb2d-d84e-4f0e-aa80-ba7f9de94ee3)
![image](https://github.com/UdaraJay/Pile/assets/25409734/b7fba6e8-6c93-4b53-941b-613bffdebd54)

((( If you want any more help with this project just lmk )))

parisosuch-dev commented:
This is great man, thank you for doing this. I've been busy and was hoping someone else took an interest in making this happen :)

UdaraJay commented:
This looks great! Thank you for taking it on :) 

I'll check this out soon and test it, if it's all good I'll publish it via the releases on this repo.

UdaraJay commented on patches/pdf-parse+1.1.1.patch:
this patch seems to be making a lot more changes than is necessary to get the package running. It should just be setting `let isDebugMode = !module.parent` to ``= false` in `index.js` and that will fix the issue.

rllyy97 replied:
Yeah I was looking at that, I haven't done any dependency patches like this before, I'll take a look at it tomorrow

GTruss commented:
@rllyy97 Thank you so much for making a PR with Windows support! I'm testing the changes now but I get the following error when I add a reply to an existing post:

usePost.js:106 Error: The "path" argument must be of type string. Received an instance of Array
    at addReplyToParent (usePost.js:113:47)
    at usePost.js:95:17
    at async index.jsx:182:7

This happens when I "Add another entry" or "Reflect".

Could this be related to the path changes in the PR? Let me know if you need more info/repro steps. Thanks again!

rllyy97 commented:
@UdaraJay I've cleaned up the patch file and fixed the issue brought up by @GTruss, should be good now.
I've also modified some of the error catching around not having an API key.  It doesn't change the experience when built at all, but the project would consistently throw errors during dev before since I didn't have a key set up.  This isn't anything specific to my changes, I noticed this behavior when I first cloned + ran the project locally. 
(I'd honestly recommend creating an application state for when the user isn't set up with the ai features so the app doesn't attempt to do them on each action)

UdaraJay commented:
Thanks @rllyy97! the patch looks good and everything works– it'll be published after some more testing on a windows machine!

The AI setup state is available from AIContext, but def a good idea to explicitly disable the reflect and embedding generation when there is no key. Your change is good for now– I'll hope to consolidate the AI APIs into one state in the future.
